django/urls/resolvers.py

```diff
@@ -107,6 +109,42 @@ def regex(self):
             self._regex_dict[language_code] = compiled_regex
         return self._regex_dict[language_code]
 
+    def describe(self):
+        """
+        Format the URL pattern for display in warning messages.
+        """
+        description = "'{}'".format(self.regex.pattern)
+        if getattr(self, 'name', False):
+            description += " [name='{}']".format(self.name)
+        return description
+
+    def _check_pattern_startswith_slash(self):
+        """
+        Check that the pattern does not begin with a forward slash.
+        """
+        regex_pattern = self.regex.pattern
+        if not settings.APPEND_SLASH:
+            # Skip check as it can be useful to start a URL pattern with a slash
+            # when APPEND_SLASH=False.
+            return []
+        if regex_pattern.startswith('/') or regex_pattern.startswith('^/'):
+            warning = Warning(
+                "Your URL pattern {} has a regex beginning with a '/'. Remove this "
+                "slash as it is unnecessary. If this pattern is targeted in an "
+                "include(), ensure the include() pattern has a trailing '/'.".format(
+                    self.describe()
+                ),
+                id="urls.W002",
+            )
+            return [warning]
+        else:
+            return []
+
+    def check(self):
```

Comment:
It sure looks like it, guess I missed the optimisation thinking there could have been other `warning.extend(*)` that should be added.